ui/src/coordinator.rs

```diff
@@ -0,0 +1,1284 @@
+            self.seqno
+                .fetch_add(1, std::sync::atomic::Ordering::Relaxed)
+        }
+
+        fn new_compile_request(&mut self) -> WSCompileRequest {
+            let seqno = self.inc_seqno();
+            WSCompileRequest {
+                target: "mir".to_owned(),
+                channel: "stable".to_owned(),
+                crate_type: "bin".to_owned(),
+                mode: "release".to_owned(),
+                edition: "2021".to_owned(),
+                tests: false,
+                backtrace: true,
+                code: r#"fn main() { println!("Hello World!"); }"#.to_owned(),
+                extra: json!({ "sequenceNumber": seqno }),
+            }
+        }
+
+        fn new_compile_request_1(&mut self) -> WSCompileRequest {
+            let seqno = self.inc_seqno();
+            WSCompileRequest {
+                target: "mir".to_owned(),
+                channel: "stable".to_owned(),
+                crate_type: "bin".to_owned(),
+                mode: "release".to_owned(),
+                edition: "2021".to_owned(),
+                tests: false,
+                backtrace: true,
+                code: r#"fn main() { println!("Hello, Compile Request The Second"); }"#.to_owned(),
+                extra: json!({ "sequenceNumber": seqno }),
+            }
+        }
+
+        fn new_format_request(&mut self) -> WSFormatRequest {
+            let seqno = self.inc_seqno();
+            WSFormatRequest {
+                code: r#"fn main() { println!("Hello World!"); }"#.to_owned(),
+                edition: "2021".to_owned(),
+                extra: json!({ "sequenceNumber": seqno }),
+            }
+        }
+
+        fn new_execute_echo_request(&mut self) -> WSExecuteRequest {
+            let seqno = self.inc_seqno();
+            WSExecuteRequest {
+                channel: "stable".to_owned(),
+                crate_type: "bin".to_owned(),
+                mode: "release".to_owned(),
+                edition: "2021".to_owned(),
+                tests: false,
+                backtrace: true,
+                code: r#"
+                    use std::io;
+
+                    fn main() -> io::Result<()> {
+                        let mut buffer = String::new();
+                        io::stdin().read_line(&mut buffer)?;
+                        print!("{}", buffer);
+                        Ok(())
+                    }
+                    "#
+                .to_owned(),
+                extra: json!({ "sequenceNumber": seqno }),
+            }
+        }
+    }
+
+    async fn check_websocket_request_response(
+        req: WSRequest,
+        timeout_millis: u64,
+    ) -> Result<WSResponse, Elapsed> {
+        let (client_sender, mut client_receiver) = spawn_worker().unwrap();
+
+        client_sender
+            .send(serde_json::to_string(&WSMessage::Request(req)).unwrap())
+            .await
+            .expect("Failed to send request to coordinator");
```

Comment:
I think that this shows that there's a bit of a impedance mismatch about where code boundaries should be. I was not expecting that we create any JSON here. The WebSocket should send JSON as that makes sense for the frontend, but decoding that should be completely independent of the coordinator and worker.

For example, we should be able to use the coordinator from a standalone command line tool that never declares serde or serde_json as a dependency (not that I have any desire to do this exact thing).

That command line tool could do something "easy" like

```rust
let result = coordinator.build_and_run_code(args).await;
```

but it could also do something "hard" like

```rust
let (worker, worker_stdin, worker_stdout, worker_stderr) = coordinator.start_build_and_run_code(args);
let output_task = tokio::spawn(async {
    // Using one task here, but maybe actually need two?
    loop {
        select! {
            Some(stdout) = worker_stdout.next() => println!("{}", stdout.into()),
            Some(stderr) = worker_stderr.next() => eprintln!("{}", stderr.into()),
            else => break,
        }
    }
});
let input_task = tokio::spawn(async {
    while let Some(stdin) = tokio::stdin().next() {
        worker_stdin.send(stdin.into());
    }
});
```

Notably, `build_and_run_code` and `start_build_and_run_code` take in normal Rust structures.

I'd also expect that `build_and_run_code` would use `start_build_and_run_code` inside the implementation by dropping stdin and gathering up stdout / stderr, then waiting for the request to finish.

Reply:
Another way of looking at this is that types like `WorkerMessage` could potentially not need to be public. For example:

```rust
// crate cw ("coordinator-worker")

mod coordinator {

    // Public
    pub struct CompileAndRunArgs;
    pub struct Coordinator;

    // Not public!
    enum MessageToDocker {
        WriteFile,
        ExecuteCommand,
    }

    impl Coordinator {
        fn start_build_and_run_code(&self, _: CompileAndRunArgs) -> (Thing1, Thing2) {
            send_message_to_docker(MessageToDocker::WriteFile);
            send_message_to_docker(MessageToDocker::ExecuteCommand);
        }
    }
}

mod worker {
    pub fn main() {}
}

// crate ui

use cw::coordinator::Coordinator;

let coord = Coordinator::new();
coord.start_build_and_run_code(...);
```

Reply:
My code boundary here are indeed wrong. I'll remove the JSON part between `coordinator` and `worker`.

To make `WorkerMessage` private, `coordinator` and `worker` has to be defined in the same crate.
In this case, we need to extract `Request`/`Response` types from `sandbox` and share them between `ui` and our new `coordinator-worker` crate.

Or we define new `Request`/`Response` types in `coordinator-worker` crate and `ui` use these new definitions to decode JSON from WebSocket? `coordinator-worker` will be a dependency of `ui` anyway.

Reply:
> Or we define new Request/Response types in coordinator-worker crate and ui use these new definitions

That would be my suggestion, yes.

> use these new definitions to decode JSON from WebSocket

But not exactly this way.

I fully admit that this is a total pain to write, but I like to have very clear lines between the layers. You can see this in the current playground code as there's both [`crate::ExecuteRequest`](https://github.com/rust-lang/rust-playground/blob/764833ed74d07e25bced2fedd4adb89dfeedebbe/ui/src/main.rs#L245-L257) (used for JSON deserialization) and [`sandbox::ExecuteRequest`](https://github.com/rust-lang/rust-playground/blob/764833ed74d07e25bced2fedd4adb89dfeedebbe/ui/src/sandbox.rs#L938-L947) (the public API to the sandbox). 

Note that the `sandbox` struct has nice domain types (like an enum for `edition`) while the `crate` struct deals with the backwards compatibility concerns inherent to the playground.

Putting logic like serde JSON deserialization into the coordinator / worker code base would be irrelevant for the hypothetical command line usage described above.

Reply:
Oh, I got that we don't have to put JSON serialization into the coordinator. I mistook requests in `sandbox` for `WSRequest`. It's `WSRequest` that deals with JSON deserialization.

> That would be my suggestion, yes.

That's to replicate `sandbox`'s request/response in the new crate, right?

Reply:
Yep, there'd probably be a `newcratename::ExecuteRequest` (and `...Response` and other actions like formatting).

Eventually, `newcratename` would completely replace the `sandbox` module of the current playground.